From 9c011b5b958059a5fa84fd725dbf2f2ba4a74c49 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Tue, 4 Mar 2025 23:02:03 +0100
Subject: [PATCH 1/3] staticd: Allocate SRv6 SIDs after connecting with SID
 Manager

On a heavily loaded system, it may take some time to establish a
connection between staticd and SRv6 SID Manager.
As a result, staticd might try to allocate SIDs in SID Manager before the
connection is fully established, leading to a failure.

This issue is reflected in the log with the following error message:

```
2025/03/04 15:24:01 STATIC: [HR66R-TWQYD][EC 100663302] srv6_manager_get_locator: invalid zclient socket
```

In this scenario, SIDs are still kept in the staticd configuration,
but not allocated and not installed in the RIB.

This commit fixes the issue by forcing staticd to retry to allocate
the SIDs once the connection with SID Manager is successfully
established.

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 staticd/static_srv6.c  | 9 +++++++++
 staticd/static_srv6.h  | 2 ++
 staticd/static_zebra.c | 2 ++
 3 files changed, 13 insertions(+)

diff --git a/staticd/static_srv6.c b/staticd/static_srv6.c
index 032bb9de9f4c..2533a37175da 100644
--- a/staticd/static_srv6.c
+++ b/staticd/static_srv6.c
@@ -153,6 +153,15 @@ void delete_static_srv6_sid(void *val)
 	static_srv6_sid_free((struct static_srv6_sid *)val);
 }
 
+void static_zebra_request_srv6_sids()
+{
+	struct static_srv6_sid *sid;
+	struct listnode *node;
+
+	for (ALL_LIST_ELEMENTS_RO(srv6_sids, node, sid))
+		static_zebra_request_srv6_sid(sid);
+}
+
 /*
  * Initialize SRv6 data structures.
  */
diff --git a/staticd/static_srv6.h b/staticd/static_srv6.h
index 48986092ae4d..2cf5f010ce95 100644
--- a/staticd/static_srv6.h
+++ b/staticd/static_srv6.h
@@ -101,6 +101,8 @@ struct static_srv6_locator *static_srv6_locator_lookup(const char *name);
 void delete_static_srv6_sid(void *val);
 void delete_static_srv6_locator(void *val);
 
+void static_zebra_request_srv6_sids(void);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/staticd/static_zebra.c b/staticd/static_zebra.c
index cc09b428365f..cc046a0e196a 100644
--- a/staticd/static_zebra.c
+++ b/staticd/static_zebra.c
@@ -187,6 +187,8 @@ static void zebra_connected(struct zclient *zclient)
 	 * in.
 	 */
 	static_install_nexthops_on_startup();
+
+	static_zebra_request_srv6_sids();
 }
 
 /* API to check whether the configured nexthop address is

From 2985bce0a60a94e74d26be59b3f65f278eab8153 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Wed, 5 Mar 2025 13:17:08 +0100
Subject: [PATCH 2/3] staticd: Fix crash when allocating SIDs without parent
 locator

In same cases, it may happen staticd attempts to allocate a SID when it
does not know about the parent locator yet.
In this case, the locator pointer stored inside the SID data structure
is NULL. When `static_zebra_request_srv6_sid` attempts to dereference
the locator pointer, we get a NULL pointer dereference crash.

This commit adds a check to prevent the crash from happening.

If the locator pointer is NULL, it means that staticd is not aware
about the parent locator yet. Then, staticd should call
`static_zebra_srv6_manager_get_locator()` to get the locator from SRv6
Manager. Once staticd receives the requested locator, the existing
callback `static_zebra_process_srv6_locator_internal()` will take care
of retrying to allocate the SID.

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 staticd/static_zebra.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/staticd/static_zebra.c b/staticd/static_zebra.c
index cc046a0e196a..a4c9c7da09e0 100644
--- a/staticd/static_zebra.c
+++ b/staticd/static_zebra.c
@@ -942,6 +942,11 @@ extern void static_zebra_request_srv6_sid(struct static_srv6_sid *sid)
 	if (!sid || !static_zebra_sid_locator_block_check(sid))
 		return;
 
+	if (!sid->locator) {
+		static_zebra_srv6_manager_get_locator(sid->locator_name);
+		return;
+	}
+
 	/* convert `srv6_endpoint_behavior_codepoint` to `seg6local_action_t` */
 	switch (sid->behavior) {
 	case SRV6_ENDPOINT_BEHAVIOR_END:

From 1cd578ff856d8e2ef917c5611470ff8eaff205b7 Mon Sep 17 00:00:00 2001
From: Carmine Scarpitta <cscarpit@cisco.com>
Date: Wed, 23 Apr 2025 06:57:57 +0000
Subject: [PATCH 3/3] staticd: Fix error reported by checkpatch

This commit fixes the following error:

```
20-Apr-2025 08:10:54	Report for static_srv6.c | 2 issues
20-Apr-2025 08:10:54	===============================================
20-Apr-2025 08:10:54	< ERROR: Bad function definition - void static_zebra_request_srv6_sids() should probably be void static_zebra_request_srv6_sids(void)
20-Apr-2025 08:10:54	< #156: FILE: /tmp/f1-3023993/static_srv6.c:156:
```

Signed-off-by: Carmine Scarpitta <cscarpit@cisco.com>
---
 staticd/static_srv6.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/staticd/static_srv6.c b/staticd/static_srv6.c
index 2533a37175da..d4d31436c1f8 100644
--- a/staticd/static_srv6.c
+++ b/staticd/static_srv6.c
@@ -153,7 +153,7 @@ void delete_static_srv6_sid(void *val)
 	static_srv6_sid_free((struct static_srv6_sid *)val);
 }
 
-void static_zebra_request_srv6_sids()
+void static_zebra_request_srv6_sids(void)
 {
 	struct static_srv6_sid *sid;
 	struct listnode *node;
